skbio/alignment/_path.py

```diff
@@ -0,0 +1,410 @@
+    states : array_like of unit8 of shape (n_segments,) or (n_packs, n_segments)
+        Packed bits representing character (0) or gap (1) status per sequence per
+        segment in the alignment.
+    starts : array_like of int of shape (n_sequences,), optional
+        Start position (0-based) of each sequence in the alignment.
+
+    See Also
+    --------
+    skbio.sequence.Sequence
+    skbio.alignment.TabularMSA
+
+    """
+
+    # comment out while working on from_cigar/to_cigar double gap problem
+    # def __str__(self):
+    #     r"""Return string representation of this AlignPath."""
+    #     return self.__repr__()
+
+    # def __repr__(self):
+    #     r"""Return summary of the alignment path."""
+    #     return (
+    #         f"<{self.__class__.__name__}, shape: {self.shape}, "
+    #         f"CIGAR: '{self.to_cigar()}'"
+    #     )
+
+    @classonlymethod
+    def from_bits(cls, bits):
+        r"""Create an alignment path from a bit array.
+
+        Parameters
+        ----------
+        bits : array_like of 0's and 1's of shape (n_seqs, alignment_length)
+            Array of zeros (char) and ones (gap) which represent the alignment.
+        """
+        # Ensure bits is a 2D array-like of ones and zeros.
+        if not isinstance(bits, np.ndarray):
+            bits = np.atleast_2d(bits)
+        if bits.ndim != 2 or bits.shape[1] == 0:
+            raise ValueError("Input 'bits' must be a non-empty 2D numpy array.")
+        if not (np.logical_or(bits == 0, bits == 1).all()):
+            raise ValueError("Input 'bits' must contain only zeros and ones.")
+
+        ints = bits[0] + (bits[1] << 1)
+        idx = np.append(0, np.where(ints[:-1] != ints[1:])[0] + 1)
+        lens = np.append(idx[1:] - idx[:-1], ints.size - idx[-1])
+        return cls(lens, ints[idx], np.zeros(bits.shape[0], dtype=int))
+
+    def to_bits(self):
+        r"""Unpack states into an array of bits."""
+        # This should be faster than the generic solution I guess.
+        # TODO: Pending benchmark and optimization.
+        if not np.all(np.isin(self.states, [0, 1, 2])):
+            raise ValueError(
+                "For pairwise alignment, `states` must only contain "
+                "zeros, ones, or twos."
+            )
+        return np.stack([self.states & 1, self.states >> 1])
+
+    def to_cigar(self, seqs=None):
+        r"""Get a CIGAR string representing the pairwise alignment path.
+
+        Parameters
+        ----------
+        seqs : list of skbio.sequence
+            Pair of aligned sequences to generate cigar string. If provided, will
+            distinguish match (=) and mismatch (X). Otherwise, will uniformly note
+            them as (mis)match (M).
+        """
+        # If `seqs` is provided, ensure that the sequence lengths match the alignment
+        # length.
+        if seqs:
+            if len(seqs[0]) != self.shape[1] or len(seqs[1]) != self.shape[1]:
```

Comment:
I am under impression that this code assumes that the two sequences are of the same length. But in fact they don't have to be. The logic of this method may be revised.